src/sciline/data_graph.py

```diff
@@ -0,0 +1,205 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
+from __future__ import annotations
+
+import itertools
+from collections.abc import Iterable
+from types import NoneType
+from typing import Any, Callable, Generator, TypeVar, Union, get_args
+
+import cyclebane as cb
+import networkx as nx
+
+from ._provider import ArgSpec, Provider, ToProvider, _bind_free_typevars
+from .handler import ErrorHandler, HandleAsBuildTimeException
+from .typing import Graph, Key
+
+
+def _as_graph(key: Key, value: Any) -> cb.Graph:
+    """Create a cyclebane.Graph with a single value."""
+    graph = nx.DiGraph()
+    graph.add_node(key, value=value)
+    return cb.Graph(graph)
+
+
+def _find_all_typevars(t: type | TypeVar) -> set[TypeVar]:
+    """Returns the set of all TypeVars in a type expression."""
+    if isinstance(t, TypeVar):
+        return {t}
+    return set(itertools.chain(*map(_find_all_typevars, get_args(t))))
+
+
+def _get_typevar_constraints(t: TypeVar) -> set[type]:
+    """Returns the set of constraints of a TypeVar."""
+    return set(t.__constraints__)
+
+
+def _mapping_to_constrained(
+    type_vars: set[TypeVar],
+) -> Generator[dict[TypeVar, type], None, None]:
+    constraints = [_get_typevar_constraints(t) for t in type_vars]
+    if any(len(c) == 0 for c in constraints):
+        raise ValueError('Typevars must have constraints')
+    for combination in itertools.product(*constraints):
+        yield dict(zip(type_vars, combination))
+
+
+T = TypeVar('T', bound='DataGraph')
+
+
+class DataGraph:
+    def __init__(self, providers: None | Iterable[ToProvider | Provider]) -> None:
+        self._cbgraph = cb.Graph(nx.DiGraph())
+        for provider in providers or []:
+            self.insert(provider)
+
+    @classmethod
+    def _from_cyclebane(cls: type[T], graph: cb.Graph) -> T:
+        out = cls([])
+        out._cbgraph = graph
+        return out
+
+    def copy(self: T) -> T:
+        return self._from_cyclebane(self._cbgraph.copy())
+
+    def __copy__(self: T) -> T:
+        return self.copy()
+
+    @property
+    def _graph(self) -> nx.DiGraph:
+        return self._cbgraph.graph
+
+    def _get_clean_node(self, key: Key) -> Any:
+        """Return node ready for setting value or provider."""
+        if key is NoneType:
+            raise ValueError('Key must not be None')
+        if key in self._graph:
+            self._graph.remove_edges_from(list(self._graph.in_edges(key)))
```

Comment:
```suggestion
            self._graph.remove_edges_from(self._graph.in_edges(key))
```
?

Reply:
I think NetworkX' `in_edges` returns a view, so that would break? @MridulS can correct me?

Reply:
Yes it's the view of the dictionary underneath, so need to create a new list object to remove the nodes.